File: OneToOneCallSample/app/src/test/java/com/stringee/apptoappcallsample/stringee/service/CallServiceOwnershipTest.java
package com.stringee.apptoappcallsample.stringee.service;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

public class CallServiceOwnershipTest {
    @Test
    public void staleSessionCannotReplaceOrStopCurrentService() {
        CallServiceOwnership ownership = new CallServiceOwnership();

        assertTrue(ownership.request(10));
        assertTrue(ownership.isRequested(10));
        assertTrue(ownership.markRunning(10));
        assertTrue(ownership.isRunning(10));
        assertFalse(ownership.request(11));
        assertFalse(ownership.release(11));
        assertTrue(ownership.isOwnedBy(10));
    }

    @Test
    public void releasedSessionAllowsTheNextSessionToOwnService() {
        CallServiceOwnership ownership = new CallServiceOwnership();

        assertTrue(ownership.request(10));
        assertTrue(ownership.markRunning(10));
        assertTrue(ownership.release(10));
        assertTrue(ownership.request(11));
        assertTrue(ownership.markRunning(11));
        assertTrue(ownership.isOwnedBy(11));
    }

    @Test
    public void markRunningRejectsAnUnrequestedGeneration() {
        CallServiceOwnership ownership = new CallServiceOwnership();

        assertFalse(ownership.markRunning(3));
        assertFalse(ownership.isOwnedBy(3));
    }
}

File: Kotlin-OneToOneCallSample/app/src/test/java/com/stringee/kotlin_onetoonecallsample/stringee/service/CallServiceOwnershipTest.kt
package com.stringee.kotlin_onetoonecallsample.stringee.service

import org.junit.Assert
import org.junit.Test

class CallServiceOwnershipTest {
    @Test
    fun staleSessionCannotReplaceOrStopCurrentService() {
        val ownership = CallServiceOwnership()

        Assert.assertTrue(ownership.request(10))
        Assert.assertTrue(ownership.isRequested(10))
        Assert.assertTrue(ownership.markRunning(10))
        Assert.assertTrue(ownership.isRunning(10))
        Assert.assertFalse(ownership.request(11))
        Assert.assertFalse(ownership.release(11))
        Assert.assertTrue(ownership.isOwnedBy(10))
    }

    @Test
    fun releasedSessionAllowsTheNextSessionToOwnService() {
        val ownership = CallServiceOwnership()

        Assert.assertTrue(ownership.request(10))
        Assert.assertTrue(ownership.markRunning(10))
        Assert.assertTrue(ownership.release(10))
        Assert.assertTrue(ownership.request(11))
        Assert.assertTrue(ownership.markRunning(11))
        Assert.assertTrue(ownership.isOwnedBy(11))
    }

    @Test
    fun markRunningRejectsAnUnrequestedGeneration() {
        val ownership = CallServiceOwnership()

        Assert.assertFalse(ownership.markRunning(3))
        Assert.assertFalse(ownership.isOwnedBy(3))
    }
}
